Weather Plugin

MarkKoz opened the pull request:
This plugin uses [Yahoo's weather API](https://developer.yahoo.com/weather/documentation.html) via the [weather-api](https://github.com/AnthonyBloomer/weather-api) Python wrapper. The plugin has two commands : `weather` and `forecast`. The former displays the current weather for a given location while the latter provides a 10-day forecast.

* Commands display weather information in an embed.
* Location is provided as a string argument. Meant to be specific e.g. a city or county.
* If weather for a location cannot be found, an error message is instead sent.
* All units default to metric. The API supports Fahrenheit/imperial units, but the plugin currently has no means of accommodating that functionality.

### Considerations
* ~~The code makes use of [f-strings](https://www.python.org/dev/peps/pep-0498/), a feature available only in Python 3.6+.~~

MarkKoz committed:
.gitignore: ignore JetBrains & Pipenv

MarkKoz committed:
weather: add weather command

This command attempts to mimic Generaloberst's command using Yahoo!'s
weather API.

MarkKoz committed:
weather: add docstrings & type annotations

MarkKoz committed:
weather command: add url & icon url to embed

* move WeatherPlugin.ICON_BASE inside the f-string

MarkKoz committed:
weather command: display high/low temps

MarkKoz committed:
weather: fix index out of range when degrees >= 360

MarkKoz committed:
weather command: add time to embed title

MarkKoz committed:
weather: add forecast command

MarkKoz committed:
forecast: add emojis for condition text

* move code thumbnail URL code to a separate function

MarkKoz committed:
weather: change get_cardinal_dir's signature to accept strings too

MarkKoz committed:
weather: fix lookup response validation

MarkKoz committed:
weather: move repeated embed code to a function

MarkKoz committed:
weather command: move temperature field to the description

* add emoji beside the condition text

MarkKoz committed:
weather: reformat embeds

* encapsulate various values in inline code blocks
* put condition and temperature on the same line for weather command

MarkKoz committed:
weather: add missing docstrings and type annotations

MarkKoz committed:
weather: replace f-strings with str.format()

Done to maintain Python 3.5 compatibility.

MarkKoz committed:
weather: revise and format docstrings

MarkKoz committed:
weather: remove type annotations

fiskenslakt commented:
Should display Fahrenheit in addition to Celsius.

fiskenslakt commented:
Set the `embed.color` attribute to the same purple as the yahoo favicon. The hex value is #7B0099 but `embed.color` expects it in base 10, so set it to `8061081`

MarkKoz commented:
While we're at it, would you prefer for it to default to Fahrenheit? For what it's worth, my preference is Celsius.

I could actually add a config option for the default.

fiskenslakt commented:
Ideally you just show both units every time, so no one is alienated when looking at the temperature.

MarkKoz commented:
I was planning on adding a command flag for it. I think it'd be too cluttered if temperature, let alone all the other values (speed and pressure), was displayed in both units. This would be especially true for the forecast.

MarkKoz committed:
weather: set embeds' colour to Yahoo! logo's purple

fiskenslakt commented:
I'm only really concerned about temperature, I doubt anyone cares about those other measurements. At the very least, I'd like it if the current temperature in the $weather command (contained in the `embed.description` IIRC) had both units. And the rest can be in just one unit, with the command flag letting you choose which one it is. 

In regards to the $forecast command, I don't think it would be too cumbersome to include both units for each day, but it's not that big of a deal. It's just something I'd like the command to have.

I won't push this issue any further than this, you make the decision and I'll merge w/e you decide to do.

MarkKoz committed:
weather: add unit parameter to commands

The unit parameter specifies if the weather data should be displayed
in imperial or metric units. It is an optional argument which defaults
to metric units. The argument is passed using the flag -u/--unit.

The valid values are 'c' and 'f', the first letters of the temperature
units associated with metric and imperial, respectively.

MarkKoz committed:
weather: add config for default weather unit

Config file is located at A2/config/weather.json.

MarkKoz committed:
weather: fix unit choices

vars was returning more than was needed. Not only is vars unnecessary,
it'd also be too much hassle to filter out its output compare to
hard-coding the unit choices.

MarkKoz committed:
weather command: display current temperature in other units too

MarkKoz commented:
An optional argument was added to specify the units. However, `$weather` will also display the _current_ temperature in the other units. I hope this is a decent compromise for now at least.

Also amended commit messages.

MarkKoz committed:
weather command: fix & format time

* fix lack of zero padding
* fix parsing of timezones that aren't exactly 3 characters in length
* format time as 12/24 hour based on unit
